### DISCUSSION TITLE
Feature comparison between `wgpu` and `DiligentEngine`.

### DISCUSSION THREAD

**Discussion by crlf0710** [gfx-rs/wgpu#2151] [General]
Hello, recently i've been using a libray called `DiligentEngine` in C++ (https://github.com/DiligentGraphics/DiligentEngine), and i'm seeking for something similar in Rust language. After some research it seems `wgpu` is the closest at this point. I'm curious what developers of wgpu consider the most important pros and cons of `wgpu` compared to that library, aside from the language difference, of course.

By the way, that library has quite a few [nice tutorial samples](https://github.com/DiligentGraphics/DiligentSamples/tree/1ee2a86231c2ceb22e88ec5bf60b2aaf1c0cf03c#tutorials). I think it would be very encourging if wgpu can provide some similar examples!

**Comment by kvark** [comment]
I interacted with Diligent just a bit. I think they are in entirely different categories: Diligent is an *engine*, while `wgpu` is a GPU abstraction (which can be a part of an engine).
So Diligent should be compared to things like [Bevy](https://bevyengine.org/) and [Rg3d](https://rg3d.rs/) instead of `wgpu`.

**Reply by crlf0710** [reply]
Thanks you for your reply. Yes, i should have wrote clearer. There's a component called DiligentCore in it that is a GPU abstraction, which can be used alone. This single component is quite similar to `wgpu` in principle.

**Reply by kvark** [reply]
Ok, I looked at it briefly, and here is what I found:
  1. DiligentCore is close to `wgpu-hal` in principle (as opposed `wgpu-core`). It's also unsafe, and also uses explicit resource transitions, although DC's Vulkan backend calls `FlushBarriers` everywhere, which I find strange. I assume it's designed to have minimal friction when implementing on different backends, but the actual details of this trade-off are likely different from WebGPU.
  2. `wgpu` (and `wgpu-hal`) follows the actual standard that's been worked on for years by the group. There is a specification and a test suite. While DC's API is more ad-hoc, and not as tested for edge cases.
  3. DC's Metal backend is not free, requires a commercial license. Other backends are the same: DX12, DX11 (which we need to revive), Vulkan, and OpenGL.
  3. DC's shader pipeline is built on SPIRV-Cross, HLSL2GLSL, glslang, DXC, or some combination of them. `wgpu-hal` uses `naga` purely, and should weigh lighter because of that.

**Reply by crlf0710** [reply]
Thanks!
